Issue 218 : Add Field version of Erf class (& Gamma)

VincentCucchietti opened the pull request:
Hey everyone, 

This pull request aims to close issue #218.

 I fieldified ```Erf``` but also ```Gamma``` as it is used in ```Erf```. I also added a ```FieldContinuedFraction``` class as it was used inside ```Gamma```. 

For the tests, I copied all existing tests and converted them to field. I also added new tests to improve code coverage.

As it is my first time fieldifying something (never thought I would say that word one day :sweat_smile: ), I would appreciate your thoughts on this and tips on how to do it correctly if I made any mistakes (as I surely did).

Cheers,
Vincent

maisonobe commented on hipparchus-core/src/main/java/org/hipparchus/util/FieldContinuedFraction.java:
```diff
@@ -0,0 +1,193 @@
+                cN = smallField;
+            }
+
+            dN = dN.pow(-1);
```
replace pow(-1) with reciprocal()

maisonobe commented on hipparchus-core/src/test/java/org/hipparchus/util/FieldContinuedFractionTest.java:
```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
```
Replace with hipparchus header.

maisonobe commented on hipparchus-core/src/test/java/org/hipparchus/util/FieldContinuedFractionTest.java:
```diff
@@ -0,0 +1,91 @@
+
+/*
+ * This is not the original file distributed by the Apache Software Foundation
+ * It has been modified by the Hipparchus project
```
Remove reference to Apache for a new file.

maisonobe commented on hipparchus-core/src/changes/changes.xml:
```diff
@@ -49,6 +49,11 @@ If the output is not quite correct, check for invisible trailing spaces!
   </properties>
   <body>
+    <release version="?" date="TBD" description="">
+      <action dev="vcucchie and luc" type="add" issue="issues/218">
```
you can remove my name here, I didn't do anything!

VincentCucchietti replied:
Now you did :smile: !

VincentCucchietti commented:
Thank you for your review @maisonobe and my bad for all these headers, everything has been fixed now :+1:

Serrof commented:
Hi Luc & Vincent,

what about an overwrite (as suggested in my comment to issue #218) of the `DerivativeStructure` version for speed?
It would go along these lines in `DSFactory `(as for `exp`, `cos`, etc.):

```    
public void erf(final double[] operand, final int operandOffset, final double[] result, final int resultOffset) {

        // create the function value and derivatives
        double[] function = new double[1 + order];
        final double const = operand[operandOffset].getReal()
        function[0] = erf(const);
        if (order > 0) {
            function[1] = (2/FastMath.sqrt(FasthMath.pi)) * FastMath.exp(-const * const);
            for (int i = 2; i <= order; ++i) {
                function[i] = - 2 * (const * function[i-1] - (i-2)* function[i];
            }
        }

        // apply function composition
        compose(operand, operandOffset, function, result, resultOffset);

}
```

Cheers,
Romain.

VincentCucchietti commented:
@BryanCazabonne

VincentCucchietti commented:
Hi @maisonobe ,

I fixed the conflict and updated the tests but the pipeline still fails with the following error: "Failed to execute goal org.sonarsource.scanner.maven:sonar-maven-plugin:3.9.1.2184:sonar" whereas i don't have any issue when building locally with maven.

As this issue concerns the sonar plugin and that i have no problem locally, i doubt that i can do something on my end, do you have an idea ? 

Configuration used: 
Apache Maven 3.6.3
java 1.8.0_362

Cheers,
Vincent

maisonobe commented:
I guess this is linked to the outdated version of Sonar.
The branch update-build-dependencies also triggers a Sonar error.
